File: leet_code/tries/trie.rb
class Trie
  def initialize()
    @root = {}
  end

  def insert(word)
    node = @root
    word.each_char do |char|
      node[char] = {} if !node[char]
      node = node[char]
    end
    node[:actual_word] = true
  end

  def search(word)
    node = @root
    word.each_char do |char|
      return false if !node[char]
      node = node[char]
    end
    return true if node[:actual_word]
    return false
  end

  def starts_with(word)
    node = @root
    word.each_char do |char|
      return false if !node[char]
      node = node[char]
    end
    true
  end

end

File: leet_code/tries/trie.js
class Trie {
    constructor() {
        this.root = {};
        this.actualWord = false;
    }

    insert(word) {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (!node[word[i]]) {
                node[word[i]] = {};
            }
            node = node[word[i]];
        }
        node.actualWord = true;


    }

    search(word) {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (!node[word[i]])return false;
            else node = node[word[i]];
        }
        if (node.actualWord) return true;
        else return false;
    }

    startsWith(word) {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (!node[word[i]]) return false;
            else node = node[word[i]];
        }
        return true;
    }
}
